### ISSUE TITLE
The accelerate execution time percentage varies greatly with the deepspeed plugin enabled or disabled

### ISSUE THREAD

**Issue by Imsovegetable** [huggingface/accelerate#3002]
I am trying to optimize the efficiency of my training scripts, and I found that there is a very big difference in the time of the same code (especially the data preparation and backward parts) when deepspeed plugin is enabled or disabled. 

My system info is as follows:
```
- `Accelerate` version: 0.32.1
- Platform: Linux-3.10.0-1160.25.1.el7_lustre.x86_64-x86_64-with-glibc2.17
- `accelerate` bash location: ~/.conda/envs/vllm/bin/accelerate
- Python version: 3.11.9
- Numpy version: 1.26.4
- PyTorch version (GPU?): 2.3.1+cu121 (True)
- PyTorch XPU available: False
- PyTorch NPU available: False
- PyTorch MLU available: False
- System RAM: 503.59 GB
- GPU type: NVIDIA GeForce RTX 4090
```

I write a toy scripts to reproduce it:
```
# test.py
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import Dataset, DataLoader
from accelerate import Accelerator
import time

# A simple dataset which generate random data
class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.size = size
        self.length = length

    def __len__(self):
        return self.length

    def __getitem__(self, index):
        return torch.randn(self.size)

# A simple neural network
class SimpleNet(nn.Module):
    def __init__(self):
        super(SimpleNet, self).__init__()
        self.fc1 = nn.Linear(1024, 8192)
        self.fc2 = nn.Linear(8192, 8192)
        self.fc3 = nn.Linear(8192, 8192)
        self.fc4 = nn.Linear(8192, 8192)
        self.fc5 = nn.Linear(8192, 1)

    def forward(self, x):
        x = x.to(torch.bfloat16)
        x = torch.relu(self.fc1(x))
        x = torch.relu(self.fc2(x))
        x = torch.relu(self.fc3(x))
        x = torch.relu(self.fc4(x))
        x = self.fc5(x)
        return x

# hyperparameters
batch_size = 4
input_size = 1024
num_workers = 32
dataset_length = 10000

# initialize everything
accelerator = Accelerator()

dataset = RandomDataset(input_size, dataset_length)
dataloader = DataLoader(dataset, batch_size=batch_size, shuffle=False, num_workers=num_workers)

model = SimpleNet().cuda()
optimizer = optim.Adam(model.parameters(), lr=0.001)
criterion = nn.MSELoss()

model, optimizer, dataloader = accelerator.prepare(model, optimizer, dataloader)

time_statistics = {
    "prepare data": 0.0,
    "forward": 0.0,
    "backward": 0.0,
    "step": 0.0,
}

model.train()
t0 = time.time()
for batch in dataloader:

    t1 = time.time()
    time_statistics["prepare data"] += t1 - t0
    t0 = t1

    # forward
    outputs = model(batch)
    loss = criterion(outputs, torch.randn(outputs.size()).to(outputs.device).to(outputs.dtype))

    t1 = time.time()
    time_statistics["forward"] += t1 - t0
    t0 = t1

    # backward
    accelerator.backward(loss)

    t1 = time.time()
    time_statistics["backward"] += t1 - t0
    t0 = t1

    # optimizer step
    optimizer.step()
    optimizer.zero_grad()

    t1 = time.time()
    time_statistics["step"] += t1 - t0
    t0 = t1

# print time statistics
if accelerator.is_local_main_process:
    print(f"Total:{sum(time_statistics.values()): .6f}")
    for k, v in time_statistics.items():
        print(f"{k}:{v: .6f}")
```
I use two config files to launch the above training with command `accelerate launch --config_file file_path test.py`. As far as I know, I think the only difference is using deepspeed or not.

The first one (named `default`) is as follows:
```
compute_environment: LOCAL MACHINE
debug: false
distributed_type: MULTI_GPU
downcast bf16: 'no'
enable_cpu_affinity: false
gpu_ids: all
machine_rank: 0
main_training_function: main
mixed precision: bf16
num machines: 1
num_processes: 8
rdzv_backend: static
same_network: true
tpu_env: []
tpu_use_cluster: false
tpu_use_sudo: false
use_cpu: false
```
And the second one (named `deepspeed`) is as follows:
```
compute_environment: LOCAL MACHINE
debug: false
deepspeed_config:
    gradient_accumulation_steps: 1
    gradient_clipping: 1.0
    offload optimizer_device: none
    offload param device: none
    zero3_init_flag: false
    zero_stage: 2
distributed type: DEEPSPEED
downcast_ bf16: 'no'
enable_cpu_affinity: false
machine rank: 0
main_ training_function: main
mixed precision: bf16
num_ machines: 1
num_processes: 8
rdzv_backend: static
same_network: true
tpu_env:[]
tpu_use_cluster: false
tpu_use_sudo: false
use_cpu: false
```
I expected that the bottleneck, or the part that takes the longest execution time, should be similar or marginly consistent for both configs. But they diverge largely on either multi gpus or single gpu situation. I run them with 8 gpus and 1 gpu seperately and collect the time statistics, which is shown below.
| Configuration           | Total time (s) | Prepare Data (s) | Forward (s) | Backward (s) | Step (s)  |
|-------------------------|-----------|------------------|-------------|--------------|-----------|
| **Default & 8 GPUs**     | 38.579844 | 36.269825        | 0.856373    | 0.903014     | 0.550632  |
| **DeepSpeed & 8 GPUs**   | 67.564035 | 0.779670         | 0.369195    | 66.414315    | 0.000855  |
| **Default & 1 GPU**      | 66.711291 | 51.200941        | 5.739730    | 6.550043     | 3.220577  |
| **DeepSpeed & 1 GPU**    | 84.181574 | 1.991175         | 2.178965    | 80.008122    | 0.003313  |

I have a few questions as follows:
1. why did it take so much time to prepare the data in `default` configuration, even num_worker has been set to 32? There doesn't seem to be any operation that takes so much time, no matter in dataset.getitem function or in dataloader.
2. why is the backward process of `deepspeed` much slower than `default`? I understand zero-2 will split gradients and optimizer into all gpus but it is still slower on single gpu setting (even x11 slower) .
3. Based on questions above, I don't know how to find the real bottleneck and how to optimize my training scripts, since the part that takes up the most time is different on different situations.

I tried to delve into source code to find out why these happpens but it is a bit difficult for me :(

Thanks!
